cpp/src/io/utilities/parsing_utils.cuh

```diff
@@ -501,33 +506,225 @@ __inline__ __device__ std::pair<char const*, char const*> trim_whitespaces_quote
 }
 
 /**
- * @brief Excludes the prefix from the input range if the string starts with the prefix.
+ * @brief Decodes a numeric value base on templated cudf type T with specified
+ * base.
  *
- * @tparam N length on the prefix, plus one
- * @param[in, out] begin Pointer to the first element of the string
- * @param end Pointer to the first element after the string
- * @param prefix String we're searching for at the start of the input range
+ * @param[in] begin Beginning of the character string
+ * @param[in] end End of the character string
+ * @param opts The global parsing behavior options
+ *
+ * @return The parsed numeric value
  */
-template <int N>
-__inline__ __device__ auto skip_if_starts_with(char const* begin,
-                                               char const* end,
-                                               const char (&prefix)[N])
+template <typename T, int base>
+__inline__ __device__ T decode_value(const char* begin,
+                                     const char* end,
+                                     parse_options_view const& opts)
 {
-  static constexpr size_t prefix_len = N - 1;
-  if (end - begin < prefix_len) return begin;
-  return thrust::equal(thrust::seq, begin, begin + prefix_len, prefix) ? begin + prefix_len : begin;
+  return cudf::io::parse_numeric<T, base>(begin, end, opts);
 }
 
 /**
- * @brief Finds the first element after the leading space characters.
+ * @brief Decodes a numeric value base on templated cudf type T
  *
- * @param begin Pointer to the first element of the string
- * @param end Pointer to the first element after the string
+ * @param[in] begin Beginning of the character string
+ * @param[in] end End of the character string
+ * @param opts The global parsing behavior options
+ *
+ * @return The parsed numeric value
  */
-__inline__ __device__ auto skip_spaces(char const* begin, char const* end)
+template <typename T,
+          std::enable_if_t<!cudf::is_timestamp<T>() and !cudf::is_duration<T>()>* = nullptr>
+__inline__ __device__ T decode_value(const char* begin,
+                                     const char* end,
+                                     parse_options_view const& opts)
+{
+  return cudf::io::parse_numeric<T>(begin, end, opts);
+}
+
+template <typename T, std::enable_if_t<cudf::is_timestamp<T>()>* = nullptr>
+__inline__ __device__ T decode_value(char const* begin,
+                                     char const* end,
+                                     parse_options_view const& opts)
+{
+  return to_timestamp<T>(begin, end, opts.dayfirst);
+}
+
+template <typename T, std::enable_if_t<cudf::is_duration<T>()>* = nullptr>
+__inline__ __device__ T decode_value(char const* begin, char const* end, parse_options_view const&)
+{
+  return to_duration<T>(begin, end);
+}
+
+// The purpose of these is merely to allow compilation ONLY
```

Comment:
I think you can make a type dispatcher type map that ignores a set of types by mapping them to `void`. That should mean that such explicit instantiations are not required. Also I would use `CUDF_UNREACHABLE` if it is possible to do so without the compiler complaining about needing a `return` statement.

See example type map here: https://github.com/rapidsai/cudf/blob/1d488098366eff2d885bce85f6de1c5e574aae7e/cpp/include/cudf/table/experimental/row_operators.cuh#L75-L79

Reply:
That's a nice solution, thanks for the suggestion. In this case, I was able to resolve the situation using `if constexpr`.